actix-router/src/resource.rs

```diff
@@ -274,15 +274,15 @@ impl ResourceDef {
             PatternType::Dynamic(ref re, ref names, len) => {
                 let mut idx = 0;
                 let mut pos = 0;
-                let mut segments: [PathItem; MAX_DYNAMIC_SEGMENTS] =
-                    [PathItem::Static(""); MAX_DYNAMIC_SEGMENTS];
+                let mut segments: [Option<PathItem>; MAX_DYNAMIC_SEGMENTS] = Default::default();
```

Comment:
Maybe we can get rid of the `Option` and the `Default` implementation by using something like:
```rust
const PATH_ITEM = PathItem::Static(Cow::Borrowed(""));
let mut segments = [PATH_ITEM; MAX_DYNAMIC_SEGMENTS];
```

Reply:
Arrays (in safe code) have to initialize all their elements, what's wrong with the `Default` implementation? Also, I guess the `Option` could be replaced with `mem::take` because `PathItem` implements `Default`. Ideally we could use `IntoIter`, but msrv is still 1.46.

Reply:
That's a trick to be able to use a non-Copy type to initialize arrays like that, all elements will be initialized.

Reply:
You don't initialize anything there.

Reply:
Yes, I know, but how is that better than `Default`?

Reply:
See:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=01f097a652dd712fd85126309e0c9abb

Reply:
I understand it works :), but why is it better than `Default`?

Reply:
To get rid of the options we need `Default` anyways.

Reply:
I guess I don't understand why you added an `Option` there.

Reply:
We can't use `IntoIter` because of the minimum supported rust version. The other (safe) way to iterate over an array by value is to iterate over mutable references, and then replace it with a default value. `Option::take` is a nice way to do this. We could use `mem::take` instead, which requires `Default`.

Reply:
Yes, sorry, I think I kinda got stuck in the array initialization and forgot about the iteration, `mem::take` seems like a better option (pun intended) though IMO.